### ISSUE TITLE
Does `denoise_audio.py` produce duration-imbalanced shards?

### ISSUE THREAD

**Issue by alephpi** [k2-fsa/OmniVoice#252]
### Checks

- [x] This template is only for usage issues encountered.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I am using English to submit this issue to facilitate community communication.

### Environment Details

Irrelevant

### Steps to Reproduce

Just run `denoise_audio.py`

### ✔️ Expected Behavior

N/A

### ❌ Actual Behavior

N/A

### Descriptions
Hi,
I'm using `denoise_audio.py` for my webdataset and I found it generate multiple shards. **However the shards are not very balanced in terms of size/audio duration, is it expected?**

I notice that the script uses `StreamLengthGroupDataset` to batch the data, and it arranges audio with similar lengths to save paddings, so I wonder does this influence the sharding such that the first shards tend to wrap-up short audios while the later shards tend to wrap-up long audios.

E.g. currently I just perform on a single tar file with 2000 .mp3 audios (~70MB), and it outputs `shard-01.tar` with 1000 .flac audios (~66MB) and `shard-02.tar` with 1000 .flac audios (~120MB), I briefly checked the `.jsonl` and it seems the second shard contain longer audio.

**Does this imbalance affect the sampling during the training?** I notice that omnivoice uses `PackingIterableDataset` for training so it seems there's no bias on the duration. But I'm probably using it for not fully Transformer model, which means the `PackingIterableDataset` with proper attention mask does not apply in my case.

**Comment by zhu-han** [comment]
Yes, this is expected with the current implementation. Output shards are split by samples_per_shard, not by total duration.

I do not think this will be an issue for training. We have both shard‑level shuffling and sample‑level shuffling during training. As a result, each training batch, which effectively comprises multiple sub‑batches under multi‑GPU training, receives samples with sufficient variation in duration.

**Comment by alephpi** [comment]
Alright. What about training on less gpus? Do you still recommend such configuration on single-GPU training for example? What might be a minimal number of GPUs that is appropriate? And also do you use the default 1000 samples per shard in omnivoice training, or do you have a better configuration on the shard size and shard number tradeoff?

In my case, I have a 2000h dataset and 1-4 gpus, I might use just a subset of it (from 200h to 2000h) to launch a training, do you have any suggestion?

**Comment by zhu-han** [comment]
Since consecutive batches will have different durations, I think this will not be a problem even under a single‑GPU setting. BTW, in many ASR/TTS training recipes, people by default adopt a length‑grouped batching strategy (e.g., StreamLengthGroupDataset in our code) to reduce computation overhead from padding. If you are also training with such a batching strategy, your final training batches will be length‑grouped anyway.

**Comment by alephpi** [comment]
Thank you for such detailed answer.

**Comment by alephpi** [comment]
For info, I run the script with different batch_size and num_workers on 2000 audio samples with an A40 , the results as follows:
```
# prefetch_factor=10
# bs=16, worker=16, 用时 3:10
# bs=8, worker=16, 用时 2:30
# bs=8, worker=8, 用时 2:30
# bs=4, worker=16, 用时 2:13
# bs=1, worker=16, 用时 2:17
# bs=1, worker=4, 用时 2:17
# bs=1, worker=1, 用时 2:16
# bs=2, worker=8, 用时 2:08
# bs=2, worker=1, 用时 2:08
# bs=4, worker=1, 用时 2:12

# prefetch_factor=2
# bs=8, worker=8, 用时 2:30
# bs=2, worker=8, 用时 2:08
# bs=2, worker=16, 用时 2:08
```
It seems enlarge the batch_size doesn't really accelerate, which is counter-intuitive but confirmed by the Sidon author, see https://github.com/sarulab-speech/Sidon/issues/16
